### PR TITLE
[WIP] - Extra features for the Node.js FFI support (such as closures)

### PR DESCRIPTION
**Pull request by danielhenrymantilla**
Branch to experiment with changes; can only be merged to #32 once that is deemed not to break the `ditto` codebase.

### CODE REVIEW

**Comment by danielhenrymantilla** [comment]
No hurries or anything, @hamchapman, but just to get somebody to somewhat review what is done here with the `Closure`s, I wouldn't mind your skimming over the current implementation to see what you think 🙂

File: src/node_js/closures.rs [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+
+            let mut ts_fun = ThreadsafeFunction::create(
+                env.raw(),
+                fun.get_thread_tied_func().unwrap(),
+                // max_queue_size /* use `0` for a sane default */
+                1,
+                // callback
+                Self::handle_params,
+            )?;
+            ts_fun.refer(&env)?; // Keep the main loop alive while this entity is.
```
Comment (hamchapman) [inline src/node_js/closures.rs:238]: The comment... does it mean "keep the main loop alive _while this entity is alive_", or is that "is" at the end about something else?

File: src/node_js/closures.rs [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+                },
+
+                | Some(func) => {
+                    // Otherwise, it means we are within the main nodejs thread,
+                    // so we can't "schedule the call and wait for it to be run",
+                    // lest we deadlock. We thus directly perform the call.
+                    //
+                    // Note: what happens if the call is done in the same thread
+                    // but from within a different `CallContext`?
+                    // Let's hope nothing bad.
```
Comment (hamchapman) [inline src/node_js/closures.rs:368]: 😁

**Review by hamchapman — APPROVED** [review]
Very nice 🙌  Awesome work getting this to play nicely 👍

File: Cargo.toml [hunk 3]
```diff
@@ -66,9 +66,24 @@ log = { version = "0.4.8", optional = true }
...
-napi = { version = "1.1.3", optional = true }
-napi-derive = { version = "1.0.0", optional = true }
 unwind_safe = "0.0.1"
+scopeguard = "1.1.0"
+
+[dependencies.napi]
+version = "1.1.3"
+optional = true
+features = ["napi5"]
+# We need `Result`-less `ThreadsafeFunction` (c.f. the associated commit),
```
[the remark this answers is not in the document]
Comment (danielhenrymantilla) [reply Cargo.toml:76]: Yeah … I haven't updated the comment sorry. Basically I have made [even more additions there](https://github.com/napi-rs/napi-rs/compare/main...getditto:ditto/closure-into-jsfunction) (which I plan to submit too to see if they get merged), and until then, we depend on our fork. Hence the submodule 🙂

File: src/node_js/closures.rs (hunk 1 above)
Comment (danielhenrymantilla) [reply src/node_js/closures.rs]: True, I've clarified that comment 🙂

File: Cargo.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (danielhenrymantilla) [reply Cargo.toml]: Clarified this too in the following commit

**Comment by danielhenrymantilla** [comment]
@hamchapman I have manually squash-merged this into https://github.com/getditto/safer_ffi/pull/32, which is the "develop" / release branch currently used by ditto. I will keep this WIP branch for some other musings 😄

**Comment by danielhenrymantilla** [comment]
@hamchapman could you skim over the last round of updates and improvements? Thanks 🙂

File: nodejs_tests/src/lib.rs [hunk 4]
```diff
@@ -78,16 +78,37 @@ fn call_with_42 (
...
+        static CB
+            : ::core::cell::Cell<
+                Option<(
+                    *mut ::std::os::raw::c_void,
+                    unsafe extern "C" fn(data: *mut ::std::os::raw::c_void, x: i32) -> u8,
+                    unsafe extern "C" fn(data: *mut ::std::os::raw::c_void),
+                )>
+            >
+            = None.into()
+        ;
```
Comment (danielhenrymantilla) [inline nodejs_tests/src/lib.rs:91]: Nevermind the "clarity" of this file, I have just used it to write potentially weird patterns that happen to cover / challenge most edge cases of the FFI.
In this instance, within this `call_with_42` callback, I am testing the three main situations where a `JsFunction` may be called:
  - same-thread, original FFI-call,
  - background thread,
  - same-thread, follow-up FFI call (hence this thread_local)

File: src/node_js/closures.rs [hunk 5]
```diff
@@ -252,6 +355,29 @@ macro_rules! impls {(
...
+        pub
+        fn make_nodejs_wait_for_this_to_be_dropped (
+            self: &'_ mut Self,
+            nodejs_should_wait: bool,
+        ) -> Result<()>
+        {
+            if nodejs_should_wait {
+                self.ts_fun.refer(&self.env)
+            } else {
+                self.ts_fun.unref(&self.env)
```
Comment (hamchapman) [inline src/node_js/closures.rs:377]: If we already `unref` by default then does this need to be here?

**Review by hamchapman — APPROVED** [review]
LGTM! Awesome work 👏

File: src/node_js/closures.rs (hunk 5 above)
Comment (danielhenrymantilla) [reply src/node_js/closures.rs:377]: I can envision somebody hitting the use case of having called `.nodejs_should_wait(true);` given a situation where they know node.js needs to wait for that closure for a while, but for some reason later on (I imagine based off some dynamic condition), it no longer being the case. They'd then need a way to "undo" that `(true)` call, which would be done by doing `.nodejs_should_wait(false); /* not anymore */`.
Does that seem like a legitimate use case? I may be doing speculative work 😅 , but it's just a boolean branch so I think it's okay 🙂

Comment (hamchapman) [reply src/node_js/closures.rs:377]: Yeah, that makes sense. 👍

**Pull request merged**